stages/org.osbuild.selinux

```diff
@@ -38,21 +45,55 @@ STAGE_OPTS = """
     "type": "array",
     "items": { "type": "string" },
     "default": ["/"]
+  },
+  "ostree": {
+    "description": "Specification for an OSTree deployment",
+    "type": "object",
+    "properties": {
+      "commit": {
+        "description": "The OSTree commit of the deployment to use",
+        "type": "string"
+      },
+      "osname": {
+        "description": "The stateroot of the operating system to use",
+        "type": "string"
+      },
+      "paths": {
+        "description": "Array paths to correct, relative to the deploy root",
+        "type": "array",
+        "items": { "type": "string" }
+      }
+    }
   }
 }
 """
 
+def setfiles(root, paths, file_contexts):
+    for path in paths:
+        print(f"SELinux: Labeling {path}")
+        subprocess.run(["setfiles", "-F",
+                        "-r", root,
+                        file_contexts,
+                        f"{root}{path}"],
+                       check=True)
+
+
 def main(tree, options):
-    file_contexts = os.path.join(f"{tree}", options["file_contexts"])
+    file_contexts = options["file_contexts"]
+    ostree = options.get("ostree")
     paths = options.get("paths", ["/"])
 
-    for path in paths:
-        subprocess.run(["setfiles",
-                        "-F",
-                        "-r", f"{tree}",
-                        f"{file_contexts}",
-                        f"{tree}"],
-                       check=True)
+    if ostree:
+        osname = ostree["osname"]
+        commit = ostree["commit"]
+        root = f"{tree}/ostree/deploy/{osname}/deploy/{commit}.0"
+
+        file_contexts = os.path.join(root, file_contexts)
+        setfiles(root, ostree.get("paths", []), file_contexts)
+    else:
+        file_contexts = os.path.join(tree, file_contexts)
+
+    setfiles(tree, paths, file_contexts)
```

Comment:
Should this `setfiles()` be in the conditional?

In general for selinux: if we do the labelling before creating the commit, is it not preserved? Does it end up wrong?

Reply:
This is intended since I am labelling the global `/boot` with it (but using the context file from the deployment) but also the `/etc` that is within the deployment:

```js
        "options": {
          "file_contexts": "etc/selinux/targeted/contexts/files/file_contexts",
          "paths": [
            "/boot"
          ],
          "ostree": {
            "commit": "711cb14fc153902b2aa75ca26a88f51269eebf4d1763bb33617b9963c6045f59",
            "osname": "fedora",
            "paths": [
              "/etc"
            ]
          }
        }
```

In general: we do level before the commit, in the `ostree.commit` stage. But when creating the image, we do a `ostree deploy`, which creates new files (three way merge of `/etc`, first and foremost, but also `/boot` because the kernel+initramfs+bls snippets get installed). Normally ostree would label those, but that internally depends on `is_selinux_enabled` being `TRUE`, which in our container it is not, we won't get proper labels. Hence we have to label in both pipelines.

Reply:
Hm, could we consider this a bug in `ostree deploy` and just do the logic they were meant to unconditionally there? Just to get the API surface "correct"?

Reply:
So I did that emulate what `ostree deploy` is doing in the stage and was happy until I realized that the `grub2` stage comes afterwards and that also 1) copies files and 2) creates new ones in `/boot/grub2/` (`grub.cfg`), 3) in `boot/efi/EFI/<vendor>/grub.cfg` as well as 4) `/etc` (`grubenv`). For 1) correctly labelling them in the build root could and then relying on `shutil.copy2` to copy the labels as well could work. For 2) we need to install `grub2-pc` in the build host copy that file over and modify it. For 3) we are out of luck, but it will end up on a FAT partition, so it might be fine. 4) is a real issue in two ways: it is in created in the totally wrong place (it should be in the deployroot) and then obviously has needs to be relabelled there.

Reply:
One idea to solve this could be to split the `org.osbuild.ostree` stage into two parts. the one that does preparses the filesystem (`ostree admin init-fs`) and then pulls (`ostree admin pull-local`) the commit (`org.osbuild.ostree.init` or `pull` ) and another one that deploys the commit (`ostree admin os-init` and `ostree admin deploy`) and have the `grub2` stage run between the two. Since in that step we fix labels in `/boot` and `deployroot/etc` this would "accidentally :("  also fix the labels of the grub2 stage. Or we keep the changes to selinux stage and run that afterwards. Or maybe better, create a new `org.osbuild.ostree.selinux` stage that does the fixups.

Reply:
Ok, ignore that idea, because I need to of course have the tree deployed for a `deployroot` to exist, so the `grub2` stage needs to come after the the `deploy`stage. :chicken: :egg:

Reply:
After various discussion, including with @martinezjavier, I introduced a new boolean, which controls whether the `/etc/default/grub` file is written at all. The idea being that no automatic tools are actually using it (it used to be grubby but that is deprecated) so it is only read if the user actively does `grub2-mkconfig -o ...`. I am not convinced that in a BLS scenario anyone would really want to run that command, instead of maybe editing `grub.cfg` directy (to tweak the timeout). Anyway, @martinezjavier and I agreed that  `/etc/grub.d/10_linux` will could be changed so that `GRUB_ENABLE_BLSCFG=true` is the default if `grubby-deprecated` is not installed (i.e. `/usr/sbin/new-kernel-pkg` not present). Then we would not need the file at all anymore, even for `grub2-mkconfig`. The only other thing we write is `GRUB_TIMEOUT` which would be reset to the default, which can easily be adjust be the user, who apparently knows what they are doing. We also don't define a bunch of other variables, so the resulting `grub.cfg` would differ from the one we wrote anyway.